src/main/java/org/elasticsearch/index/translog/fs/FsTranslog.java

```diff
@@ -359,70 +431,81 @@ public Location add(Operation operation) throws TranslogException {
                 current.sync();
             }
 
-            assert new BytesArray(current.read(location)).equals(bytes);
-
-            FsTranslogFile trans = this.trans;
-            if (trans != null) {
-                try {
-                    location = trans.add(bytes);
-                } catch (ClosedChannelException e) {
-                    // ignore
-                }
-            }
-            Releasables.close(bytes);
-            released = true;
+            assert current.assertBytesAtLocation(location, bytes);
             return location;
         } catch (Throwable e) {
             throw new TranslogException(shardId, "Failed to write operation [" + operation + "]", e);
         } finally {
-            rwl.readLock().unlock();
-            if (!released && out != null) {
+            if (out != null) {
                 Releasables.close(out.bytes());
             }
         }
     }
 
     @Override
-    public FsChannelSnapshot snapshot() throws TranslogException {
-        while (true) {
-            FsChannelSnapshot snapshot = current.snapshot();
-            if (snapshot != null) {
-                return snapshot;
+    public Snapshot snapshot() {
+        try (ReleasableLock _ = readLock.acquire()) {
+            final FsChannelReader[] readers = uncomittedTranslogs.toArray(new FsChannelReader[uncomittedTranslogs.size() + 1]);
+            readers[readers.length - 1] = current;
+            return createdSnapshot(readers);
+        }
+    }
+
+    private Snapshot createdSnapshot(FsChannelReader... translogs) {
+        ArrayList<FsChannelSnapshot> channelSnapshots = new ArrayList<>();
+        boolean success = false;
+        try {
+            for (FsChannelReader translog : translogs) {
+                channelSnapshots.add(translog.snapshot());
+            }
+            Snapshot snapshot = new FsTranslogSnapshot(channelSnapshots, logger);
+            success = true;
+            return snapshot;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(channelSnapshots);
             }
-            Thread.yield();
         }
     }
 
     @Override
-    public Snapshot snapshot(Snapshot snapshot) {
-        FsChannelSnapshot snap = snapshot();
-        if (snap.translogId() == snapshot.translogId()) {
-            snap.seekTo(snapshot.position());
+    public Translog.View view() {
```

Comment:
I think this should be named `newView`, otherwise it's ambiguous whether it returns a new or a current view